locator/topology.cc

```diff
@@ -565,6 +565,26 @@ void topology::for_each_node(std::function<void(const node*)> func) const {
     }
 }
 
+std::unordered_set<const node*> topology::get_nodes() const {
+    std::unordered_set<const node*> nodes;
+    for (const auto& np : _nodes) {
+        if (np && !np->left() && !np->is_none()) {
```

Comment:
When can `np` be a `nullptr`?

Reply:
```cpp
node_holder topology::pop_node(const node* node) {
    tlogger.trace("topology[{}]: pop_node: {}, at {}", fmt::ptr(this), node_printer(node), lazy_backtrace());

    unindex_node(node);

    auto nh = std::exchange(_nodes[node->idx()], {});

    // shrink _nodes if the last node is popped
    // like when failing to index a newly added node
    if (std::cmp_equal(node->idx(), _nodes.size() - 1)) {
        _nodes.resize(node->idx());
    }

    return nh;
}
```

`_nodes` is a vector and not a set. When we remove a node from topology, we exchange its entry in `_nodes` with `nullptr`. We only remove the trailing `nullptr` entries.